storage/scratchpad/src/sparse_merkle/sparse_merkle_test.rs

```diff
@@ -549,3 +575,189 @@ fn test_drop() {
+    let mut smt_serial;
+    let mut smt_batch;
+    for i in 1..=5 {
+        for _ in 0..1000 {
+            let key = HashValue::random_with_rng(&mut rng);
+            let value = AccountStateBlob::from(HashValue::random_with_rng(&mut rng).to_vec());
+            kvs.push((key, value));
+        }
+        let batch = kvs.iter().map(|(k, v)| (*k, v)).collect::<Vec<_>>();
+        let start = std::time::Instant::now();
+        smt_serial = smt
+            .serial_update(vec![batch.clone()], &proof_reader)
+            .unwrap()
+            .1;
+        println!("serial {}-th run: {}ms", i, start.elapsed().as_millis());
+
+        let start = std::time::Instant::now();
+        smt_batch = smt.batch_update(batch.clone(), &proof_reader).unwrap();
+        println!("batch {}-th run: {}ms", i, start.elapsed().as_millis());
+        kvs.clear();
+        assert_eq!(smt_serial.root_hash(), smt_batch.root_hash());
+    }
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig::with_cases(50))]
+    #[test]
+    fn test_read_partial_tree_from_storage(input in arb_blocks_to_commit().no_shrink()) {
+        let tmp_dir = TempPath::new();
+        let db = DiemDB::new_for_test(&tmp_dir);
+        let mut cur_ver = 0;
+        for (txns_to_commit, ledger_info_with_sigs) in input.iter() {
+            db.save_transactions(
+                &txns_to_commit,
+                cur_ver, /* first_version */
+                Some(ledger_info_with_sigs),
+            )
+            .unwrap();
+            cur_ver += txns_to_commit.len() as u64;
+            let (latest_version ,state_root_hash) = db.get_latest_state_root().unwrap();
+            prop_assert_eq!(latest_version + 1, cur_ver);
+            let updates = (0..100)
+                .into_iter()
+                .map(|_| {
+                    (
+                        AccountAddress::random(),
+                        AccountStateBlob::from(HashValue::random().to_vec()),
+                    )
+                }).collect::<Vec<_>>();
+            let account_state_proofs = updates
+                .iter().map(|(k, _)| {
+                    db.get_account_state_with_proof(*k, latest_version, latest_version)
+                        .map(|p| p.proof.transaction_info_to_account_proof().clone())
+                })
+                .collect::<Result<Vec<_>>>()
+                .unwrap();
+            let proof_reader = ProofReader::new(
+                itertools::zip_eq(
+                    updates.iter().map(|(k, _)| k.hash()),
+                    account_state_proofs.into_iter(),
+                ).collect::<Vec<_>>(),
+            );
+            let smt = SparseMerkleTree::new(state_root_hash);
+            let input = updates.iter().map(|(k, v)| (k.hash(), v)).collect::<Vec<_>>();
+            let batch_updated_tree = smt.batch_update(input.clone(), &proof_reader).unwrap();
+            let serial_updated_tree = smt.serial_update(vec![input], &proof_reader).unwrap().1;
```

Comment:
I'd call `serial_update()` on `updates` (per version batches) here to cover more branches in your code.

Reply:
K.. I mean, `updates` better be a vec<vec<>> so we can cover the case where batch_update() is called on a smt with a non-unknown root.